File: 48-rotate-image/48-rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    let n = matrix.length -1;
    let s = (n+1)/2;
    for(let i=0;i<s;i++){
        for(let j = i ; j<n-i ; j++){
            let x = i, y = j,t = matrix[x][y];
            for(let k=1;k<=4;k++){
                let z = matrix[y][n-x];
                x = n-x;
                matrix[y][x] = t;
                t = z;
                [x,y] = [y,x];
            }
        }
    }
};

File: 48-rotate-image/48-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& m) {
        int n = m.size()-1;
        int s = (n+1)/2;
        for(int i=0;i<s;i++){
            for(int j=i;j<n-i;j++){
                int t = m[i][j];
                int x = i,y = j;
                for(int k=1;k<=4;k++){
                    int z = m[y][n-x];
                    x = n-x;
                    m[y][x] = t;
                    t= z;swap(y,x);
                }
            }
        }
    }
};